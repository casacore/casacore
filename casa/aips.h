//# aips.h: Global initialization for standard types, etc.
//# Copyright (C) 1993-1999,2000,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_AIPS_H
#define CASA_AIPS_H

//# Define compiler specific flags
#include <casacore/casa/aipsenv.h>

//# Define the standard types used by Casacore.
#include <casacore/casa/aipstype.h>

//# Define the extra non-standard types used by Casacore
#include <casacore/casa/aipsxtype.h>

//# Define the special Casacore macros
#include <casacore/casa/aipsdef.h>

namespace casa { //# NAMESPACE CASA - BEGIN

} //# NAMESPACE CASA - END

#endif
