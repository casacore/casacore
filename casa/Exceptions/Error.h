//# Error.h: Base class for all Casacore errors
//# Copyright (C) 1993,1994,1995,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_ERROR_H
#define CASA_ERROR_H



#include <sys/types.h>
#include <casa/aips.h>
#include <casa/BasicSL/String.h>
#include <casa/OS/Mutex.h>
#include <exception>


namespace casa { //# NAMESPACE CASA - BEGIN

// Throw the given exception with a string composed of various arguments.
// E.g.
// <srcblock>
//    CASATHROW (AipsError, "integer=" << myint << ", float=" << myfloat);
// </srcblock>
#define CASATHROW(exc, arg) do {     \
    std::ostringstream casa_log_oss; \
    casa_log_oss << arg;             \
    throw exc(casa_log_oss.str());   \
  } while (0)

// Throw an AipsError exception containing the file and line.
#define ThrowCc(m) throw AipsError ((m), __FILE__, __LINE__)

// Throw an AipsError exception if the condition is true.
#define ThrowIf(c,m) {if (c) {casa::AipsError::throwIf (True, (m), __FILE__, __LINE__, __PRETTY_FUNCTION__);}}

// Throw an AipsError exception if the system error code is not 0.
// It adds the message for that error code to the exception text.
#define ThrowIfError(c,m) {if (c) {casa::AipsError::throwIfError (True, (m), __FILE__, __LINE__, __PRETTY_FUNCTION__);}}

// Repackage and rethrow an AipsError exception.
#define Rethrow(e,m) {throw casa::AipsError::repackageAipsError ((e),(m),__FILE__,__LINE__, __PRETTY_FUNCTION__);}


// <summary>Base class for all Casacore library errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <prerequisite>
//   <li> ExcpError
// </prerequisite>
//
// <synopsis>
//  This is the base class for all of the Casacore error classes. Because
//  all of the errors have a common base class, any error can be caught
//  with a single catch statement.
//
//  This class has a string which allows error messages to be propagated.
//
//  <note role=tip> The string member must be handled very carefully because
//        string is also derived from cleanup, thus the 
//        <src>message.makePermanent()</src> call in the implementation of
//        the constructors. This prevents the String from being cleaned up
//        in the middle of an exception.
//  </note>
//
// </synopsis>
//
// <example>
// <srcblock>
//      throw(AipsError("SOME STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class AipsError: public std::exception
{
public:

  enum Category {
    BOUNDARY, INITIALIZATION, INVALID_ARGUMENT, CONFORMANCE,
    ENVIRONMENT, SYSTEM, PERMISSION, GENERAL
  };

  //
  // Simply returns the stored error message.
  //
  virtual const char* what() const throw()
    { return(message.c_str()); }
  const String &getMesg() const
    { return(message); }
  String getStackTrace() const;
  AipsError::Category getCategory( ) const
    { return(category); }

  // Append a message. This is used by LogIO when an exception is logged.
  // The message is const to be able to use it for a temporary exception.
  void setMessage (const String& msg) const
    { const_cast<AipsError*>(this)->message = msg; }

  // Creates an AipsError and initializes the error message from
  // the parameter.
  // <group>
  AipsError (const Char *str, Category c = GENERAL);
  AipsError (const String &str, Category c = GENERAL);
  AipsError (const String &msg, const String &filename, uInt lineNumber,
             Category c = GENERAL);
  AipsError (Category c = GENERAL);
  // </group>

  //
  // Destructor which does nothing.
  //
  ~AipsError() throw();

  // Get or clear the stacktrace info.
  // <group>
  static void getLastInfo (String & message, String & stackTrace);
  static String getLastMessage ();
  static String getLastStackTrace ();
  static void clearLastInfo ();
  // </group>

  // Repackage an exception.
  static AipsError repackageAipsError (AipsError& error, 
                                       const String& message,
                                       const char* file,
                                       Int line,
                                       const char* func);

  // Throw if the condition is true.
  static void throwIf (Bool condition, const String& message,
                       const char* file, Int line,
                       const char* func = "");

  // Throw if the system error code is not 0.
  static void throwIfError (Int errorCode, const String& prefix,
                            const char* file, Int line,
                            const char* func = "");

protected:
  // Add the stack trace to the message (if USE_STACKTRACE is set).
  void addStackTrace ();

  String message;
  Category category;
  String stackTrace;
};


// <summary>Allocation errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
//
// This class is used for allocation errors. It adds an extra
// data item, the failed allocation size. Otherwise much the
// same as <src>AipsError</src>.
//
// </synopsis>
//
// <example>
// <srcblock>
//     throw(AllocError("ANY STRING",1024));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class AllocError : public AipsError {
protected:
  size_t Size;
public:
  //
  // This constructor takes the error message and the failed
  // allocation size.
  //
  // <group>
  AllocError(const Char *str, uInt sze) : AipsError(str,SYSTEM), Size(sze) {}
  AllocError(const String &str, uInt sze) : AipsError(str,SYSTEM), Size(sze)  {}
  // </group>

  //
  // This function returns the failed allocation size.
  //
  size_t size() const {return(Size);}

  //
  // Destructor which does nothing.
  //
  ~AllocError() throw();

};


// <summary>Base class for all indexing errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is the base class of all <src>IndexError</src>s. It is
// defined to allow the user to catch any of the many kinds of IndexErrors
// which may be thrown. It can also be thrown itself if returning
// the illegal index value is unimportant.
// </synopsis>
//
// <example>
// <srcblock>
//     throw(IndexError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class IndexError : public AipsError {
public:
  //
  // Creates an GeneralIndexError and initializes the error message from
  // the parameter
  // <group>
  IndexError(const Char *str,Category c=BOUNDARY) : AipsError(str,c) {}
  IndexError(const String &str,Category c=BOUNDARY) : AipsError(str,c) {}
  IndexError(Category c=BOUNDARY) : AipsError(c) {}
  // </group>

  //
  // Destructor which does nothing.
  //
  ~IndexError() throw();
};


// <summary>Index errors returning the bad index</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is templated to allow generalalized indexes to be returned
// with the error message i.e. the class is templated on the index type.
//
// </synopsis>
//
// <example>
// <srcblock>
//     throw(indexError<int>(3,"ANY STRING"));/
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

template<class t> class indexError : public IndexError {
protected:
  t oIndex;                 // Offending Index
public:
  //
  // This constructor takes the error message and the index
  // which cause the error to occur.
  //
  // <group>
  indexError(t oI, const Char *str, Category c=BOUNDARY);
  indexError(t oI, const String &str, Category c=BOUNDARY);
  indexError(t oI, Category c=BOUNDARY) : IndexError(c), oIndex(oI) {};
  // </group>

  //
  // Destructor which does nothing.
  //
  ~indexError() throw();
};


// <summary>Duplicate key errors</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This class is the base class of all duplicate key errors. It is
// defined to allow the user to catch any of the many kinds of DuplErrors
// which may be thrown. It can also be thrown itself if returning
// the illegal key is unimportant.
// </synopsis>
//
// <example>
// <srcblock>
//    throw(DuplError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class DuplError : public AipsError {
public:
  //
  // Creates an DuplError and initializes the error message from
  // the parameter
  // <group>
  DuplError(Category c=BOUNDARY) : AipsError(c) {}
  DuplError(const Char *str,Category c=BOUNDARY) : AipsError(str,c) {}
  DuplError(const String &str,Category c=BOUNDARY) : AipsError(str,c) {}
  // </group>

  //
  // Destructor which does nothing.
  //
  ~DuplError() throw();
};


// <summary>Duplicate key errors where the bad key is returned</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
//  This template is for generalized duplicate key errors where the template
//  type parameter is the type of the key which caused the error. Because this
//  class is derived from <linkto class=DuplError><src>DuplError</src>
//  </linkto>, the user to catch all duplicate key errors with one catch
//  statement. 
//
// </synopsis>
//
// <example>
//     throw(duplError<int>(4,"ANY STRING"));
// </example>
//
// <todo asof="">
// </todo>

template<class t> class duplError : public DuplError {
protected:
  t oKey;                   // Offending Key
public:
  //
  // This constructs a "duplError" for the offending key, and an
  // optional character string.
  //
  // <group>
  duplError(t oI, const Char *str,Category c=BOUNDARY);
  duplError(t oI, const String &str,Category c=BOUNDARY);
  duplError(t oI,Category c=BOUNDARY) : DuplError(c), oKey(oI) {};
  // </group>

  //
  // Destructor which does nothing.
  //
  ~duplError() throw();
};


// <summary>Exception for an error in a system call</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This error is to be used for if a system call returns an error.
// It uses strerror to get the system error message.
// </synopsis>

class SystemCallError : public AipsError
{
public:
  // This constructs a "SystemCallError" from the system call function name
  // and the errno.
  SystemCallError(const String &funcName, int error, Category c=GENERAL);

  SystemCallError (int error, const String &msg, const String &filename,
                   uInt lineNumber, Category c=GENERAL);

  // Destructor which does nothing.
  ~SystemCallError() throw();

  // Get the errno.
  int error() const
    { return itsError; }

  // Get the message belonging to an error.
  static String errorMessage(int error);

private:
  int itsError;
};


// <summary>Exception which halts execution</summary>
// <use visibility=export>
//
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// <synopsis>
// This error causes an execution to halt regardless. It
// causes execution to halt before the exception can be caught.
// </synopsis>
//
// <example>
// <srcblock>
//     throw(AbortError("ANY STRING"));
// </srcblock>
// </example>
//
// <todo asof="">
// </todo>

class AbortError : public AipsError {
public:
  //
  // This constructs a "AbortError" from the error message.
  //
  // <group>
  AbortError(const Char *str,Category c=GENERAL);
  AbortError(const String &str,Category c=GENERAL);
  // </group>

  //
  // Destructor which does nothing.
  //
  ~AbortError() throw();
};



} //# NAMESPACE CASA - END

#ifdef AIPS_NEEDS_RETHROW
#ifndef CASACORE_NEEDS_RETHROW
#define CASACORE_NEEDS_RETHROW
#endif
#endif

#ifdef CASACORE_NEEDS_RETHROW
#define RETHROW(X) throw(X);
#else
#define RETHROW(X)
#endif

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casa/Exceptions/Error.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
