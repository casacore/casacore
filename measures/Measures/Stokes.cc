//# Stokes.cc: Stokes parameter definitions for interface to table data
//# Copyright (C) 1994,1995,1997,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/aips.h>
#include <measures/Measures/Stokes.h>
#include <casa/BasicSL/String.h>

namespace casa { //# NAMESPACE CASA - BEGIN

Stokes::StokesTypes Stokes::type(Int stokesNumber) 
{
   StokesTypes val = Undefined;
   if (stokesNumber > Undefined && stokesNumber < NumberOfTypes) {
      val = StokesTypes(stokesNumber);
   }
   return val;
}

Stokes::StokesTypes Stokes::type(const String &stokesName) 
{
   StokesTypes val = Undefined;
   String name = stokesName;
   name.upcase();
   if      (name == "I")  val = I;
   else if (name == "Q")  val = Q;
   else if (name == "U")  val = U;
   else if (name == "V")  val = V;
   else if (name == "RR") val = RR;
   else if (name == "RL") val = RL;
   else if (name == "LR") val = LR;
   else if (name == "LL") val = LL;
   else if (name == "XX") val = XX;
   else if (name == "XY") val = XY;
   else if (name == "YX") val = YX;
   else if (name == "YY") val = YY;
   else if (name == "RX") val = RX;
   else if (name == "RY") val = RY;
   else if (name == "LX") val = LX;
   else if (name == "LY") val = LY;
   else if (name == "XR") val = XR;
   else if (name == "XL") val = XL;
   else if (name == "YR") val = YR;
   else if (name == "YL") val = YL;
   else if (name == "PP") val = PP;
   else if (name == "PQ") val = PQ;
   else if (name == "QP") val = QP;
   else if (name == "QQ") val = QQ;
   else if (name == "RCIRCULAR") val = RCircular;
   else if (name == "LCIRCULAR") val = LCircular;
   else if (name == "LINEAR") val = Linear;
   else if (name == "PTOTAL") val = Ptotal;
   else if (name == "PLINEAR") val = Plinear;
   else if (name == "PFTOTAL") val = PFtotal;
   else if (name == "PFLINEAR") val = PFlinear;
   else if (name == "PANGLE") val = Pangle;
   return val;
}

String Stokes::name(StokesTypes stokesType) 
{
   String stokesName;
   switch (stokesType) {
   case Undefined: stokesName="??"; break;
   case I: stokesName="I"; break;
   case Q: stokesName="Q"; break;
   case U: stokesName="U"; break;
   case V: stokesName="V"; break;
   case RR: stokesName="RR"; break;
   case RL: stokesName="RL"; break;
   case LR: stokesName="LR"; break;
   case LL: stokesName="LL"; break;
   case XX: stokesName="XX"; break;
   case XY: stokesName="XY"; break;
   case YX: stokesName="YX"; break;
   case YY: stokesName="YY"; break;
   case RX: stokesName="RX"; break;
   case RY: stokesName="RY"; break;
   case LX: stokesName="LX"; break;
   case LY: stokesName="LY"; break;
   case XR: stokesName="XR"; break;
   case XL: stokesName="XL"; break;
   case YR: stokesName="YR"; break;
   case YL: stokesName="YL"; break;
   case PP: stokesName="PP"; break;
   case PQ: stokesName="PQ"; break;
   case QP: stokesName="QP"; break;
   case QQ: stokesName="QQ"; break;
   case RCircular: stokesName="RCircular"; break;
   case LCircular: stokesName="LCircular"; break;
   case Linear: stokesName="Linear"; break;
   case Ptotal: stokesName="Ptotal"; break;
   case Plinear: stokesName="Plinear"; break;
   case PFtotal: stokesName="PFtotal"; break;
   case PFlinear: stokesName="PFlinear"; break;
   case Pangle: stokesName="Pangle"; break;
   }
   return stokesName;
}

Fallible<Int> Stokes::receptor1(StokesTypes stokesType)
{
    Int rec1 = (stokesType-1)%4; 
    if (rec1<2) rec1=0; else rec1=1;
    if (stokesType>Stokes::V && stokesType<Stokes::RCircular) 
	return Fallible<Int>(rec1);
    else return Fallible<Int>();
}

Fallible<Int> Stokes::receptor2(StokesTypes stokesType)
{
    Int rec2 = (stokesType-1)%4; 
    if (rec2==0 || rec2==2) rec2=0; else rec2=1;
    if (stokesType>Stokes::V && stokesType<Stokes::RCircular) 
	return Fallible<Int>(rec2);
    else return Fallible<Int>();
}

Int Stokes::FITSValue(StokesTypes which)
{
   Int retval;

   switch (which) {
   case I: retval = 1; break;
   case Q: retval = 2; break;
   case U: retval = 3; break;
   case V: retval = 4; break;
   case RR: retval = -1; break;
   case LL: retval = -2; break;
   case RL: retval = -3; break;
   case LR: retval = -4; break;
   case XX: retval = -5; break;
   case YY: retval = -6; break;
   case XY: retval = -7; break;
   case YX: retval = -8; break;
   case PFlinear: retval = 6; break;
   case Pangle: retval = 7; break;
   default: retval = 100 + Int(which);
   }
   return retval;
}

Stokes::StokesTypes Stokes::fromFITSValue(Int which)
{
   StokesTypes retval;

   switch (which) {
   case 1: retval = I; break;
   case 2: retval = Q; break;
   case 3: retval = U; break;
   case 4: retval = V; break;
   case -1: retval = RR; break;
   case -2: retval = LL; break;
   case -3: retval = RL; break;
   case -4: retval = LR; break;
   case -5: retval = XX; break;
   case -6: retval = YY; break;
   case -7: retval = XY; break;
   case -8: retval = YX; break;
   case 6: retval = PFlinear; break;
   case 7: retval = Pangle; break;
   default: retval = Stokes::type(which-100);
   }
   return retval;
}

} //# NAMESPACE CASA - END

