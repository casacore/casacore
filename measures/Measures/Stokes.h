//# Stokes.h: Stokes parameter definitions for interface to table data
//# Copyright (C) 1994,1995,1996,1997,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef MEASURES_STOKES_H
#define MEASURES_STOKES_H

#include <casa/aips.h>
#include <casa/BasicSL/String.h>
#include <casa/Utilities/Fallible.h>
#include <casa/Arrays/Vector.h>

namespace casa { //# NAMESPACE CASA - BEGIN

// <summary>
// Stokes parameter definitions for interface to table data.
// </summary>

// <reviewed tests="tStokes">
// </reviewed>

// <synopsis>
// This enumerates the available Stokes types, but does not define
// the operations for conversion between Stokes types.
// This class is a wrapper for the ENUM and conversion functions.
// </synopsis>

class Stokes {
public:

// The Stokes types are defined by this enum.
//
// <note role=warning>
// <b>DO NOT CHANGE THE ORDER OF THESE TYPES</b>, as the integers corresponding
// to the <src>enum</src> are required for storage in Tables.
// One can add to these types, but the order must be preserved.
// The correlation products are required to have the order indicated with
// values of 1,2,3,4 plus n*4.
// </note>
//
//# The enum comments below are placed in the position they are to make the
//# extracted documentation look nice.
	enum StokesTypes { 
	    // undefined value = 0
	    Undefined=0, 
	    I,  
	    Q,  
	    U,  
	    // standard stokes parameters 
	    V,
	    //
	    RR, 
	    RL, 
	    LR, 
	    // circular correlation products
	    LL,
	    //
	    XX, 
	    XY, 
	    YX, 
	    // linear correlation products
	    YY,
	    //
	    RX, 
	    RY, 
	    LX, 
	    LY,
	    XR, 
	    XL, 
	    YR, 
	    // mixed correlation products
	    YL,
	    //
	    PP, 
	    PQ, 
	    QP, 
	    // general quasi-orthogonal correlation products
	    QQ, 
	    //
	    RCircular, 
	    LCircular, 
	    // single dish polarization types
	    Linear,
	    // Polarized intensity ((Q^2+U^2+V^2)^(1/2))
	    Ptotal,
	    // Linearly Polarized intensity ((Q^2+U^2)^(1/2))
	    Plinear,
	    // Polarization Fraction (Ptotal/I)
	    PFtotal,
	    // Linear Polarization Fraction (Plinear/I)
	    PFlinear,
	    // Linear Polarization Angle (0.5 arctan(U/Q)) (in radians)
	    Pangle
	};

//      The number of StokesTypes.
// <note role=warning>
// <b>Update</b> <src>NumberOfTypes</src> when entries are added.
// </note>
	enum {
	    // The number of StokesTypes.
	    NumberOfTypes = 33
	    };
	
//		convert Int to StokesTypes, returns Stokes::Undefined if 
//		it is an invalid type
	static StokesTypes type(Int stokesNumber);

//		convert String to StokesTypes, returns Stokes::Undefined if
//		it is an unrecognized string.  The valid strings are the
//		same as the characters used in the enum above (i.e.
//		"I" returns Stokes::I, "Linear" returns Stokes::Linear, etc).
        static StokesTypes type(const String & stokesName);

//		convert StokesTypes to String, Stokes::Undefined returns
//		"??".
	static String name(StokesTypes stokesType);

	// get all recognized stokes names in no guaranteed order.
	static Vector<String> allNames(Bool includeUndefined = False);

//              map StokesTypes to receptor number (0 or 1) for the
//              interferometric correlation products.
//              e.g. XY will give receptor1==0 receptor2==1 etc.
//              I,Q,U,V and the single dish types will produce invalid
//              Fallible.
//              <group>
	static Fallible<Int> receptor1(StokesTypes stokesType);
	static Fallible<Int> receptor2(StokesTypes stokesType);
//              </group>

    // These two functions map stokes type to FITS type and vice versa. If you add a 
    // StokesType you should change these functions as well.
    // <ul>
    //     <li> I,Q,U,V <-> 1,2,3,4
    //     <li> RR,LL,RL,LR <-> -1,-2,-3,-4   Note! Not the same as enum order!
    //     <li> XX,YY,XY,YX <-> -5,-6,-7,-8   Note! Not the same as enum order!
    //     <li> Otherwise, FITS type <-> 100 + Int(stokesType). This is not standard FITS.
    // </ul>
    // <group>
        static Int    FITSValue(StokesTypes which);
        static StokesTypes fromFITSValue(Int);
    // </group>
private:
};


} //# NAMESPACE CASA - END

#endif

