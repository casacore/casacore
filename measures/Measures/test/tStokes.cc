//# tStokes.cc: This program tests Stokes interface class to table data
//# Copyright (C) 1994,1995,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes

#include <casa/aips.h>
#include <measures/Measures/Stokes.h>
#include <casa/BasicSL/String.h>
#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
int main() {

	try {
		Int polint;
		String polstr;

		polint=5;
		cout << "polint = " << polint
				<< ", Stokes::type(polint) = " << Stokes::type(polint)
		<<",  Stokes::type(Stokes::type(polint)) = "
		<< Stokes::type(Stokes::type(polint))
		<< endl
		<<  " --- receptor1 = "<<Stokes::receptor1(Stokes::type(polint))
		<<  " --- receptor2 = "<<Stokes::receptor2(Stokes::type(polint))
		<< endl;

		polint=-1;
		try {
			cout << "polint = " << polint;
			cout << ", Stokes::type(polint) = " << Stokes::type(polint);
			cout << ", Stokes::type(Stokes::type(polint)) = ";
			cout << Stokes::type(Stokes::type(polint));
			cout << endl;
			cout <<  " --- receptor1 = ";
			cout << Stokes::receptor1(Stokes::type(polint));
			cout <<  " --- receptor2 = ";
			cout << Stokes::receptor2(Stokes::type(polint));
			cout << endl;
		} catch(AipsError x) {
			cout << " Caught exception of receptor correctly: "<<x.getMesg()<<endl;
		}

		polstr="XY";
		cout << "polstr = " << polstr
				<< ", Stokes::type(polstr) = " << Stokes::type(polstr)
		<< ", Stokes::type(Stokes::type(polstr)) = "
		<< Stokes::type(Stokes::type(polstr))
		<< endl
		<<  " --- receptor1 = "<<Stokes::receptor1(Stokes::type(polstr))
		<<  " --- receptor2 = "<<Stokes::receptor2(Stokes::type(polstr))
		<< endl;

		polstr="LX";
		cout << "polstr = " << polstr
				<< ", Stokes::type(polstr) = " << Stokes::type(polstr)
		<< ", Stokes::type(Stokes::type(polstr)) = "
		<< Stokes::type(Stokes::type(polstr))
		<< endl
		<<  " --- receptor1 = "<<Stokes::receptor1(Stokes::type(polstr))
		<<  " --- receptor2 = "<<Stokes::receptor2(Stokes::type(polstr))
		<< endl;


		try {
			polstr="AB";
			cout << "polstr = " << polstr;
			cout << ", Stokes::type(polstr) = " << Stokes::type(polstr);
			cout << ", Stokes::type(Stokes::type(polstr)) = ";
			cout << Stokes::type(Stokes::type(polstr));
			cout << endl  ;
			cout <<  " --- receptor1 = ";
			cout << Stokes::receptor1(Stokes::type(polstr)) ;
			cout <<  " --- receptor2 = ";
			cout << Stokes::receptor2(Stokes::type(polstr)) ;
			cout << endl;
		} catch(AipsError x) {
			cout << " Caught exception of receptor correctly: "<<x.getMesg()<<endl;
		}
		for (uInt i=0;i<Stokes::NumberOfTypes;i++) {
			if (Stokes::fromFITSValue(Stokes::FITSValue(Stokes::type(i)))
			!= Stokes::type(i)) {
				cerr << "Stokes FITS value conversion failed" << endl;
				cerr << "  FITSValue(" << i << ") = " << Stokes::FITSValue(Stokes::type(i)) << endl;
				cerr << "  fromFITSValue of that value = "
						<< Stokes::fromFITSValue(Stokes::FITSValue(Stokes::type(i))) << endl;
			}
		}
		Vector<String> allNames = Stokes::allNames();
		cout << "All names: ";
		for (uInt i=0; i<allNames.size(); i++) {
			cout << allNames[i] << " ";
		}
		cout << endl;
		AlwaysAssert(Stokes::allNames(False).size() == Stokes::NumberOfTypes - 1, AipsError);
		AlwaysAssert(Stokes::allNames(True).size() == Stokes::NumberOfTypes, AipsError);
		cout << "ok" << endl;
	}
	catch (AipsError) {
		cout << "fail" << endl;

	}
	return 0;
}
