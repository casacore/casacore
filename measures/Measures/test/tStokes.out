polint = 5, Stokes::type(polint) = 5,  Stokes::type(Stokes::type(polint)) = 5
 --- receptor1 = 0 --- receptor2 = 0
polint = -1, Stokes::type(polint) = 0, Stokes::type(Stokes::type(polint)) = 0
 --- receptor1 =  Caught exception of receptor correctly: Fallible<T>:: invalid object accessed. Sorry I don't know from where
polstr = XY, Stokes::type(polstr) = 10, Stokes::type(Stokes::type(polstr)) = 10
 --- receptor1 = 0 --- receptor2 = 1
polstr = LX, Stokes::type(polstr) = 15, Stokes::type(Stokes::type(polstr)) = 15
 --- receptor1 = 1 --- receptor2 = 0
polstr = AB, Stokes::type(polstr) = 0, Stokes::type(Stokes::type(polstr)) = 0
 --- receptor1 =  Caught exception of receptor correctly: Fallible<T>:: invalid object accessed. Sorry I don't know from where
